src/ui/mainwindowwithdetachabledockwidgets.cc

```diff
@@ -0,0 +1,717 @@
+      if (main_window != nullptr) {
+        auto* dock_widget = dynamic_cast<DockWidget*>(
+            main_window->tabToDockWidget(tab_bar, tab_bar->currentIndex()));
+        if (dock_widget != nullptr) {
+          MainWindowWithDetachableDockWidgets::setActiveDockWidget(dock_widget);
+        }
+      }
+    }
+  }
+}
+
+bool MainWindowWithDetachableDockWidgets::event(QEvent* event) {
+  bool has_children = !findChildren<QDockWidget*>().empty();
+
+  if (has_children) {
+    if (event->type() == QEvent::ChildAdded) {
+      emit childAdded(static_cast<QChildEvent*>(event)->child());
+    } else if (event->type() == QEvent::ChildRemoved) {
+      emit childRemoved();
+    }
+  } else {
+    if (event->type() == QEvent::ChildRemoved && this != first_main_window_) {
+      this->deleteLater();
+    }
+  }
+
+  return QMainWindow::event(event);
+}
+
+bool MainWindowWithDetachableDockWidgets::splitDockWidgetImpl(
+    QMainWindow* main_window, QDockWidget* first, QDockWidget* second,
+    Qt::Orientation orientation) {
+  // As QMainWindow::splitDockWidget doesn't handle multiple dock widgets
+  // tabified together, it's necessary to detach all excessive docks,
+  // split dock area and then reattach the docks.
+  auto tabified_docks = main_window->tabifiedDockWidgets(first);
+  for (auto dock_widget : tabified_docks) {
+    main_window->removeDockWidget(dock_widget);
+  }
+  tabified_docks.removeOne(second);
+
+  second->show();
+  main_window->splitDockWidget(first, second, orientation);
+
+  for (auto dock_widget : tabified_docks) {
+    dock_widget->show();
+    main_window->tabifyDockWidget(first, dock_widget);
+  }
+
+  return !main_window->tabifiedDockWidgets(first).contains(second);
+}
+
+void MainWindowWithDetachableDockWidgets::createHexEditTab(
+    const QSharedPointer<FileBlobModel>& data_model) {
+  QSharedPointer<QItemSelectionModel> selection_model(
+      new QItemSelectionModel(data_model.data()));
+
+  auto* node_widget = new NodeWidget(this, data_model, selection_model);
+  addTab(node_widget, data_model->path().join(" : "), nullptr);
+}
+
+void MainWindowWithDetachableDockWidgets::createHexEditTab(
+    const QString& fileName, const dbif::ObjectHandle& fileBlob) {
+  QSharedPointer<FileBlobModel> data_model(
+      new FileBlobModel(fileBlob, {QFileInfo(fileName).fileName()}));
+  createHexEditTab(data_model);
+}
+
+void MainWindowWithDetachableDockWidgets::createVisualization(
+    const QSharedPointer<FileBlobModel>& data_model) {
+  auto* panel = new visualization::VisualizationPanel(this, data_model);
+
+  panel->setData(
+      QByteArray(reinterpret_cast<const char*>(data_model->binData().rawData()),
+                 static_cast<int>(data_model->binData().size())));
+  panel->setAttribute(Qt::WA_DeleteOnClose);
+
+  // FIXME: main_window_ needs to be updated when docks are moved around,
+  // then we can use this behaviour without any weird effects
+  // auto sibling = DockWidget::getParentDockWidget(this);
```

Comment:
Hmm, you've removed this FIXME from another place, if it doesn't make sense anymore please also remove it from here ;)

Reply:
Actually it might make sense here. I hope to actually fix it in next PR

Reply:
Well ok, actually it KIND OF make sense. The commented code is invalid, yet I would like to leave a mention about actually passing a pointer to sibling here. I love focus on this in one of my next PRs

Reply:
Sure, let's deal with it in another PR.